src/Reflection/ClassReflection.php

```diff
@@ -114,21 +133,50 @@ public function getFileName()
 	 */
 	public function getParentClass()
 	{
-		if ($this->reflection->getParentClass() === false) {
+		$parentClass = $this->reflection->getParentClass();
+
+		if ($parentClass === false) {
 			return false;
 		}
 
-		return $this->broker->getClass($this->reflection->getParentClass()->getName());
+		$extendsTag = $this->getFirstExtendsTag();
+
+		if ($extendsTag !== null && $this->isValidAncestorType($extendsTag->getType(), [$parentClass->getName()])) {
+			$type = $extendsTag->getType();
+
+			if ($this->isGeneric()) {
+				$type = TemplateTypeHelper::resolveTemplateTypes(
+					$type,
+					$this->getActiveTemplateTypeMap()
+				);
+			}
+
+			if (!$type instanceof GenericObjectType) {
+				return $this->broker->getClass($parentClass->getName());
+			}
+
+			return $type->getClassReflection() ?? $this->broker->getClass($parentClass->getName());
+		}
+
+		return $this->broker->getClass($parentClass->getName());
 	}
 
 	public function getName(): string
 	{
 		return $this->reflection->getName();
 	}
 
-	public function getDisplayName(): string
+	public function getDisplayName(bool $withTemplateTypes = true): string
 	{
-		return $this->displayName;
+		$name = $this->displayName;
+
+		if ($withTemplateTypes === false || $this->resolvedTemplateTypeMap === null) {
+			return $name;
+		}
+
+		return $name . '<' . implode(',', array_map(static function (Type $type): string {
+			return $type->describe(VerbosityLevel::typeOnly());
+		}, $this->resolvedTemplateTypeMap->getTypes())) . '>';
```

Comment:
It is useful to have this by default: Rules that print class names will automatically print the template types. For example: `Parameter #1 $a of method A<int>::set() expects int, string given.`

Reply:
Is there any case when we would want ` $withTemplateTypes` with false?

Reply:
I want to use `getDisplayName()` in generic type, while keeping some control over how template types are described.